src/coreclr/jit/objectalloc.cpp

```diff
@@ -1335,3 +1482,1699 @@ void ObjectAllocator::RewriteUses()
+
+    // Must be an equality or inequality
+    //
+    if (!tree->OperIs(GT_NE, GT_EQ))
+    {
+        JITDUMP("... not NE/EQ\n");
+        return nullptr;
+    }
+
+    GenTree* op1     = tree->gtGetOp1();
+    GenTree* op2     = tree->gtGetOp2();
+    bool     swapped = false;
+
+    // gdv creates NE(hnd, indir(locl))
+    // but let's not rely on that
+    //
+    if (!op1->OperIs(GT_IND))
+    {
+        swapped = true;
+        std::swap(op1, op2);
+    }
+
+    if (!op1->OperIs(GT_IND))
+    {
+        JITDUMP("... no JTRUE(cmp(ind, ...))\n");
+        return nullptr;
+    }
+
+    if (!op1->TypeIs(TYP_I_IMPL))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int, ...))\n");
+        return nullptr;
+    }
+
+    GenTree* const addr = op1->AsIndir()->Addr();
+
+    if (!addr->TypeIs(TYP_REF))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(*:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!addr->OperIs(GT_LCL_VAR))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!op2->IsIconHandle(GTF_ICON_CLASS_HDL))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), clsHnd))\n");
+        return nullptr;
+    }
+
+    // Passed the checks... fill in the info.
+    //
+    info->m_local  = addr->AsLclVar()->GetLclNum();
+    bool isNonNull = false;
+    bool isExact   = false;
+    info->m_type   = (CORINFO_CLASS_HANDLE)op2->AsIntCon()->gtIconVal;
+
+    JITDUMP("... " FMT_BB " is guard for V%02u\n", block->bbNum, info->m_local);
+    return tree;
+}
+
+//------------------------------------------------------------------------------
+// CheckForGuardedUse - see if this use of lclNum is controlled by a failing
+//    GDV check that we're tracking as part of conditional escape.
+//
+// Arguments:
+//    block  - block containing tree
+//    tree   - parent tree using the local
+//    lclNum - local being read
+//
+// Returns:
+//    true if this use is a conditionally escaping use.
+//
+bool ObjectAllocator::CheckForGuardedUse(BasicBlock* block, GenTree* tree, unsigned lclNum)
+{
+    // Find pseudo local...
+    //
+    unsigned pseudoLocal = BAD_VAR_NUM;
+    if (!m_EnumeratorLocalToPseudoLocalMap.TryGetValue(lclNum, &pseudoLocal))
+    {
+        JITDUMP("... no pseudo local?\n");
+        return false;
+    }
+
+    // Verify that this call is made under a **failing** GDV test
+    //
+    GuardInfo info;
+    if (!IsGuarded(block, tree, &info, /* testOutcome */ false))
+    {
+        JITDUMP("... not guarded?\n");
+        return false;
+    }
+
+    // Find the GDV guard for the pseudo-local
+    //
+    CloneInfo* pseudoGuardInfo;
+    if (!m_CloneMap.Lookup(pseudoLocal, &pseudoGuardInfo))
+    {
+        JITDUMP("... under non-gdv guard?\n");
+        return false;
+    }
+
+    // Verify this appearance is under the same guard
+    //
+    if ((info.m_local == lclNum) && (pseudoGuardInfo->m_local == lclNum) && (info.m_type == pseudoGuardInfo->m_type))
```

Comment:
Does the `m_type == pseudoGuardInfo->m_type` check work correctly for NAOT? From what I can tell one is an actual compile time class handle from `GenTreeAllocObj::gtAllocObjClsHnd`, while the other is the constant that we were pointing out is not a compile time handle above. Perhaps that one needs to be changed to use `GenTreeIntCon::gtCompileTimeHandle`.

Reply:
Could be -- I will take a look.

Reply:
that's why I pointed out that we should never cast IntCon to CLASS_HANDLE and use a helper instead (feel free to rename it)🙂. I suspect it shouldn't be too hard to eliminate compile time handle for classes just like we don't need it in many other handle types

Reply:
Yes, it should use the compile time handle from the icon. It is currently hard to reach this opt in R2R/NAOT because it requires GDV and hence PGO.

I'll have a PR up in a bit with this and some of the other deferred fixes.

Reply:
I guess for R2R/NAOT the priority may not be that high given that now we have late devirt inlining.